egs2/TEMPLATE/asr1/scripts/utils/evaluate_asr.sh

```diff
@@ -0,0 +1,266 @@
+    --nj             # Number of parallel jobs (default="${nj}").
+    --gpu_inference  # Whether to use gpu in the inference (default="${gpu_inference}").
+
+    # Model related configuration
+    --model_tag       # Model tag or url available in espnet_model_zoo (default="${model_tag}")
+                      # If provided, overwrite --asr_model_file and --lm_file options.
+    --asr_model_file  # ASR model file path in local (default="${asr_model_file}").
+    --lm_file         # LM model file path in local (default="${lm_file}").
+
+    # Inference related configuration
+    --inference_config  # ASR inference configuration file (default="${inference_config}").
+    --inference_args    # Additional arguments for ASR inference (default=${inference_args}).
+
+    # Scoring related configuration
+    --bpemodel    # BPE model path, needed if you want to calculate TER (default="${bpemodel}").
+    --nlsyms_txt  # Non-language symbol file (default="${nlsyms_txt}").
+    --cleaner     # Text cleaner module for the reference (default="${cleaner}").
+    --gt_text     # Kaldi-format groundtruth text file (default="${gt_text}")
+                  # This must be provided if you want to calculate scores.
+
+Examples:
+    # Use pretrained model and perform only inference
+    $0 --model_tag <model_tag> wav.scp asr_outputs
+
+    # Use pretrained model and perform inference and scoring
+    $0 --model_tag <model_tag> --stop-stage 2 --gt_text /path/to/text wav.scp asr_results
+
+    # Use local model and perform inference and scoring
+    $0 --asr_model_file /path/to/model.pth --stop-stage 2 --gt_text /path/to/text wav.scp asr_results
+
+EOF
+)
+
+log "$0 $*"
+# shellcheck disable=SC1091
+. utils/parse_options.sh
+
+wavscp=$1
+outdir=$2
+
+if [ $# -ne 2 ]; then
+    log "${help_message}"
+    exit 2
+fi
+
+# shellcheck disable=SC1091
+. ./path.sh
+# shellcheck disable=SC1091
+. ./cmd.sh
+
+# Check the option is valid
+if [ -z "${gt_text}" ] && [ "${stop_stage}" -ge 2 ]; then
+    log "--gt_text must be provided if perform scoring. Skipped scoring."
+    exit 1
+fi
+if [ -z "${model_tag}" ] && [ -z "${asr_model_file}" ]; then
+    log "Either --model_tag or --asr_model_file must be provided."
+    exit 1
+fi
+
+if ${gpu_inference}; then
+    # shellcheck disable=SC2154
+    _cmd="${cuda_cmd}"
+    _ngpu=1
+else
+    # shellcheck disable=SC2154
+    _cmd="${decode_cmd}"
+    _ngpu=0
+fi
+
+if [ ${stage} -le 1 ] && [ ${stop_stage} -ge 1 ]; then
+    _opts=()
+    if [ -n "${inference_config}" ]; then
+        _opts+=("--config")
+        _opts+=("${inference_config}")
+    fi
+    if [ -n "${asr_model_file}" ]; then
+        _opts+=("--asr_model_file")
+        _opts+=("${asr_model_file}")
+    fi
+    if [ -n "${lm_file}" ]; then
+        _opts+=("--lm_file")
+        _opts+=("${lm_file}")
+    fi
+    if [ -n "${model_tag}" ]; then
+        # FIXME: workaround until fixing filelock
+        espnet_model_zoo_download --unpack true "${model_tag}"
```

Comment:
Would you consider the support for using the configuration packed inside the model for inference and decoding? e.g. the language model, BPE configs.
> In [egs2/TEMPLATE/asr1/asr.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/asr1/asr.sh#L1350-L1374), it allows packing `lm_file`, `bpemodel`, `nlsyms_txt` and other files along with `asr_model_file` into the model.

I used to try something like the following to parse the BPE config from the downloaded model:
```bash
log "Downloading ${asr_model} for decoding and evaluation"
asr_exp="${download_dir}/${asr_model// /_}"
mkdir -p "${asr_exp}"

# If the model already exists, you can skip downloading
espnet_model_zoo_download --unpack true "${asr_model}" > "${asr_exp}/config.txt"

# Get the path of each file
_asr_model_file=$(<"${asr_exp}/config.txt" sed -e "s/.*'asr_model_file': '\([^']*\)'.*$/\1/")
_asr_train_config=$(<"${asr_exp}/config.txt" sed -e "s/.*'asr_train_config': '\([^']*\)'.*$/\1/")

# Get token_type
token_type=$(<"${_asr_train_config}" grep -Eo "^token_type: (\w+)$" | awk '{print $2}')
if [ "${token_type}" = "bpe" ]; then
    bpemodel=$(<"${_asr_train_config}" grep -Eo "^bpemodel: .*$" | awk '{print $2}' | sed -e 's/^"\(.*\)"/\1/' -e "s/^'\(.*\)'/\1/")
    if [ -z "${bpemodel}" ]; then
        log "Invalid model: token_type=bpe, but 'bpemodel' is not defined in '${_asr_train_config}'"
        exit 1
    elif [ ! -f "${bpemodel}" ]; then
        log "No such file for 'bpemodel': ${bpemodel}"
        exit 1
    fi
    # nbpe_mode=$(echo "${bpemodel}" | grep -Po "(?<=/bpe_)\w+(?=/bpe\.model$)")
    # nbpe=$(echo "${nbpe_mode}" | grep -Po "\d+")
    # bpemode=$(echo "${nbpe_mode}" | grep -Po "[a-zA-Z]+")
else
    bpemodel=none
fi

# Create symbolic links
ln -sf "${_asr_model_file}" "${asr_exp}"
ln -sf "${_asr_train_config}" "${asr_exp}"
inference_asr_model=$(basename "${_asr_model_file}")

if [ "$(<${asr_exp}/config.txt grep -c lm_file)" -gt 0 ]; then
    _lm_file=$(<"${asr_exp}/config.txt" sed -e "s/.*'lm_file': '\([^']*\)'.*$/\1/")
    _lm_train_config=$(<"${asr_exp}/config.txt" sed -e "s/.*'lm_train_config': '\([^']*\)'.*$/\1/")

    lm_exp="${expdir}/${download_model// /_}/lm"
    mkdir -p "${lm_exp}"

    ln -sf "${_lm_file}" "${lm_exp}"
    ln -sf "${_lm_train_config}" "${lm_exp}"
    inference_lm=$(basename "${_lm_file}")
fi
```

Reply:
If we use model_tag, the model will be loaded through `from_pretrained` function.
https://github.com/espnet/espnet/blob/bd23667ed1fb34e62d2a2c89030001840feeb61f/espnet2/bin/asr_inference.py#L262-L275
So the lm or bpe setting will be loaded automatically.
And this download part is needed to prevent from simultaneous download when using parallel `asr_inference` jobs (basically, this can prevent with `filelock` but now espnet_model_zoo's file lock does not work well).

Reply:
I see. It very cool to have the support already!